pkg/controller/stats/controller.go

```diff
@@ -0,0 +1,227 @@
+			COUNT(*) as TotalCols
+		FROM cluster('{cluster}', INFORMATION_SCHEMA.COLUMNS)
+		WHERE table_catalog == 'default'
+		GROUP BY table_name, table_catalog, Shard
+		) as t2
+		ON t1.DatabaseName = t2.DatabaseName and t1.TableName = t2.TableName and t1.Shard = t2.Shard`,
+	// average writing rate for all tables per second
+	insertRateQuery: `
+SELECT
+	sd.Shard,
+	sd.RowsPerSecond,
+	sd.BytesPerSecond
+FROM (
+	SELECT
+		shardNum() as Shard,
+		(intDiv(toUInt32(date_trunc('minute', toDateTime(event_time))), 2) * 2) * 1000 as t,
+		TRUNCATE(avg(ProfileEvent_InsertedRows),0) as RowsPerSecond,
+		formatReadableSize(avg(ProfileEvent_InsertedBytes)) as BytesPerSecond,
+		ROW_NUMBER() OVER(PARTITION BY shardNum() ORDER BY t DESC) rowNumber
+	FROM cluster('{cluster}', system.metric_log)
+	GROUP BY t, shardNum()
+	ORDER BY t DESC, shardNum()
+	) sd
+WHERE sd.rowNumber=1`,
+	stackTracesQuery: `
+SELECT
+	shardNum() as Shard,
+	arrayStringConcat(arrayMap(x -> demangle(addressToSymbol(x)), trace), '\\n') AS trace_function,
+	count()
+FROM cluster('{cluster}', system.stack_trace)
+GROUP BY trace_function, Shard
+ORDER BY count()
+DESC SETTINGS allow_introspection_functions=1`,
+}
+
+type StatsController struct {
+	kubeClient        kubernetes.Interface
+	clickhouseConnect *sql.DB
+}
+
+func NewStatsController(
+	kubeClient kubernetes.Interface,
+) *StatsController {
+	c := &StatsController{
+		kubeClient: kubeClient,
+	}
+	return c
+}
+
+func (c *StatsController) GetDiskInfo(namespace string) ([][]string, error) {
+	data, err := c.getDataFromClickHouse(diskQuery, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("error when getting diskInfo from clickhouse: %v", err)
+	}
+	return data, nil
+}
+
+func (c *StatsController) GetTableInfo(namespace string) ([][]string, error) {
+	data, err := c.getDataFromClickHouse(tableInfoQuery, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("error when getting tableInfo from clickhouse: %v", err)
+	}
+	return data, nil
+}
+
+func (c *StatsController) GetInsertRate(namespace string) ([][]string, error) {
+	data, err := c.getDataFromClickHouse(insertRateQuery, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("error when getting insertRate from clickhouse: %v", err)
+	}
+	return data, nil
+}
+
+func (c *StatsController) GetStackTraces(namespace string) ([][]string, error) {
+	data, err := c.getDataFromClickHouse(stackTracesQuery, namespace)
+	if err != nil {
+		return nil, fmt.Errorf("error when getting stackTraces from clickhouse: %v", err)
+	}
+	return data, nil
+}
+
+func (c *StatsController) getDataFromClickHouse(query int, namespace string) ([][]string, error) {
+	var err error
+	if c.clickhouseConnect == nil {
+		c.clickhouseConnect, err = util.SetupClickHouseConnection(c.kubeClient, namespace)
+		if err != nil {
+			return nil, err
+		}
+	}
+	result, err := c.clickhouseConnect.Query(queryMap[query])
+	if err != nil {
+		return nil, fmt.Errorf("failed to get data from clickhouse: %v", err)
```

Comment:
do we want to reset db connection in case certain connection error is encountered, so that upon next query we can retry establishing a new conn?